src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Google LLC.
+//
+// Author: Eddy Hsu <eddyhsu@google.com>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/data_blob.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <rtos/init.h>
+
+#include <google_ctc_audio_processing.h>
+#include <google_ctc_audio_processing_sof_message_reader.h>
+
+LOG_MODULE_REGISTER(google_ctc_audio_processing, CONFIG_SOF_LOG_LEVEL);
+
+/* bf0e1bbc-dc6a-45fe-bc90-2554cb137ab4 */
+DECLARE_SOF_RT_UUID("google-ctc-audio-processing", google_ctc_audio_processing_uuid,
+		    0xbf0e1bbc, 0xdc6a, 0x45fe, 0xbc, 0x90, 0x25, 0x54, 0xcb,
+		    0x13, 0x7a, 0xb4);
+
+DECLARE_TR_CTX(google_ctc_audio_processing_tr, SOF_UUID(google_ctc_audio_processing_uuid),
+	       LOG_LEVEL_INFO);
+
+struct google_ctc_audio_processing_comp_data;
+
+typedef void (*ctc_func)(struct google_ctc_audio_processing_comp_data *cd,
+			 const struct audio_stream *source,
+			 struct audio_stream *sink,
+			 struct input_stream_buffer *input_buffers,
+			 struct output_stream_buffer *output_buffers,
+			 uint32_t frames);
+
+struct ctc_func_map {
+	enum sof_ipc_frame fmt; /**< source frame format */
+	ctc_func func; /**< processing function */
+};
+
+struct google_ctc_audio_processing_comp_data {
+	float *input;
+	float *output;
+	uint32_t input_ptr;
+	uint32_t output_ptr;
+	uint32_t num_frames;
+	uint32_t num_channels;
+	GoogleCtcAudioProcessingState *state;
+	struct comp_data_blob_handler *tuning_handler;
+	bool reconfigure;
+	ctc_func ctc_func;
+};
+
+static inline int16_t convert_float_to_int16(float data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	xtfloat x1;
+	int16_t x;
+
+	x1 = XT_MUL_S(x0, ratio);
+	x = XT_TRUNC_S(x1, 0);
```

Comment:
FWIW, the use of intrinsics for scalar float usage is minimally useful.  The compiler generates great (generally identical) code if you just use C types, and has the advantage of working correctly on HIFI4/5 devices without all the hardware muckery.  The code in the AEC patch is solving basically identical problems and generating more or less optimal output code with only floats.

Also the !HIFI case (which corresponds to "open source toolchain" in practice) is sort of pessimal.  Q_CONVERT_FLOAT() is going to go through a double precision conversion, where if you just let gcc see the floats directly it'll call into its own libgcc soft-float routines at single precision.

Reply:
Changed to the same version as AEC.